server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -1182,41 +1156,27 @@ public void evaluateOnDemandTabletsForUnload() {
 
     final SortedMap<KeyExtent,Tablet> online = getOnlineTablets();
 
-    // Find and remove access time entries for KeyExtents
-    // that are no longer in the onlineTablets collection
-    Set<KeyExtent> missing = onDemandTabletAccessTimes.keySet().stream()
-        .filter(k -> !online.containsKey(k)).collect(Collectors.toSet());
-    if (!missing.isEmpty()) {
-      log.debug("Removing onDemandAccessTimes for tablets as tablets no longer online: {}",
-          missing);
-      missing.forEach(onDemandTabletAccessTimes::remove);
-      if (onDemandTabletAccessTimes.isEmpty()) {
-        return;
-      }
-    }
-
-    // It's possible, from a tablet split or merge for example,
-    // that there is an on-demand tablet that is hosted for which
-    // we have no access time. Add any missing online on-demand
-    // tablets
-    online.forEach((k, v) -> {
-      if (v.isOnDemand() && !onDemandTabletAccessTimes.containsKey(k)) {
-        insertOnDemandAccessTime(k);
+    // Sort the extents so that we can process them by table.
+    final SortedMap<KeyExtent,Long> sortedOnDemandExtents = new TreeMap<>();
+    // We only want to operate on OnDemand Tablets
+    online.entrySet().forEach((e) -> {
+      if (e.getValue().isOnDemand()) {
+        sortedOnDemandExtents.put(e.getKey(), e.getValue().getLastAccessTime());
       }
     });
 
-    log.debug("Evaluating online on-demand tablets: {}", onDemandTabletAccessTimes);
-
-    if (onDemandTabletAccessTimes.isEmpty()) {
+    if (sortedOnDemandExtents.isEmpty()) {
       return;
     }
 
+    log.debug("Evaluating online on-demand tablets: {}", sortedOnDemandExtents);
+
     // If the TabletServer is running low on memory, don't call the SPI
     // plugin to evaluate which on-demand tablets to unload, just get the
     // on-demand tablet with the oldest access time and unload it.
     if (getContext().getLowMemoryDetector().isRunningLowOnMemory()) {
       final SortedMap<Long,KeyExtent> timeSortedOnDemandExtents = new TreeMap<>();
-      onDemandTabletAccessTimes.forEach((k, v) -> timeSortedOnDemandExtents.put(v.get(), k));
+      sortedOnDemandExtents.forEach((k, v) -> timeSortedOnDemandExtents.put(v, k));
       Long oldestAccessTime = timeSortedOnDemandExtents.firstKey();
```

Comment:
This is an existing issue, but could fix it here.   Think this is sorting on nano time which is not correct because nano time has an arbitrary starting point in a process and it can wrap.  So should find the largest difference.

```suggestion
      long currTime = System.nanoTime();
      sortedOnDemandExtents.forEach((k, v) -> timeSortedOnDemandExtents.put(v-currTime, k));
      Long oldestAccessTime = timeSortedOnDemandExtents.lasttKey();
```